### DISCUSSION TITLE
Upsert operator

### DISCUSSION THREAD

**Discussion by jacoscaz** [afs/rdf-delta#159] [Q&A]
Hi all,

I’ve been looking at the RDF patch format for use within an experimental IoT project based on RDF. The format looks very promising but, in the interest of conciseness, I was looking for something that would allow me to “upsert” a triple with a different object term without separate `A` and `D` entries. 

For example, given a store with `<Alice> foaf:knows <Bob> .`, I’d be looking for something like 

```
U <Alice> foaf:knows <Peter> .
```

to replace 

```
D <Alice> foaf:knows <Bob> .
A <Alice> foaf:knows <Peter> . 
```

Does anything like that exist? If so, has it ever been considered within RDF patch? Are there other approaches I should be looking into?

**Comment by afs** [comment]
It's not so clear cut - suppose 

```
<Alice> foaf:knows <Bob> .
<Alice> foaf:knows <Charlie> .
```
how does the upsert know which, or both, triples is the intended target?

This is what SPARQL Update comes in - the WHERE clause can include conditions, FILTERs. 

RDF patch is the record of changes. It can even be made reversible if the application tests, when generating the patch, whether `D` or `A` operations have an effect.

There is LDpatch but it only operates on  RDF of a specific shape (LDP data is a tree if using blank nodes).

[SPARQLPatch](https://www.w3.org/2001/sw/wiki/SparqlPatch) is a subset of SPARQL Update to make implementation lighter.
